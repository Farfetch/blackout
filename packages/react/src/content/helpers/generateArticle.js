import * as schemaProperties from '../components/schemas/schemaProperties';
import { getCategories, getMetatag, stripUrlSubfolder } from '../utils';

/**
 * Generate Structured Data (JSON-LD) for an Article.
 *
 * @function generateArticle
 * @memberof module:content/helpers
 *
 * @param {object} metadata - All SEO metadata for the Article.
 * @param {object} date - All information about Article date.
 * @param {string} date.publicationDate - Publication date of the Article.
 * @param {string} date.modificationDate - Modification date of the Article.
 * @param {string} url - Relative url of the page (location.pathname).
 * @param {string} title - Title of the first component in the Page (e.g. Hero Title, or H1).
 * @param {string} image - Image URL to describe the content of the Page.
 * @param {string} author - Author name of the Article.
 * @param {Array}  breadcrumbs - Breadcrumbs list with all url Texts.
 * @param {object} publisher - Object with publisher data.
 * @param {string} publisher.name - The name of the Article publisher.
 * @param {string} publisher.url - The url of the Article publisher.
 * @param {string} publisher.logo - The logo path of the Article publisher.
 *
 * @returns {object} - JSON-LD Schema.org object for an Article.
 */
export default (
  metadata,
  date,
  url,
  title,
  image,
  author,
  breadcrumbs,
  publisher,
) => {
  const generateSchemaOrgProperty = property =>
    getMetatag(property, metadata?.metatags);

  return {
    '@context': schemaProperties.DEFAULT_CONTEXT,
    '@type': schemaProperties.DEFAULT_TYPE,
    name: generateSchemaOrgProperty('og:title') || metadata?.title,
    headline: metadata?.h1 || title,
    description:
      generateSchemaOrgProperty('og:description') || metadata?.description,
    url: generateSchemaOrgProperty('og:url') || stripUrlSubfolder(url),
    mainEntityOfPage: {
      '@type': schemaProperties.DEFAULT_TYPE_WEBPAGE,
      '@id': url,
    },
    datePublished:
      generateSchemaOrgProperty('article:datePublished') ||
      date?.publicationDate,
    dateModified:
      generateSchemaOrgProperty('article:dateModified') ||
      date?.modificationDate,
    publisher: {
      '@type': schemaProperties.DEFAULT_ORGANISATION,
      name:
        generateSchemaOrgProperty('article:publisher:name') || publisher?.name,
      url: generateSchemaOrgProperty('article:publisher:url') || publisher?.url,
      logo: {
        '@type': schemaProperties.DEFAULT_IMAGE_OBJECT,
        url:
          generateSchemaOrgProperty('article:publisher:logo') ||
          publisher?.logo,
      },
    },
    author: {
      '@type': schemaProperties.DEFAULT_THING,
      name: generateSchemaOrgProperty('og:site_name') || author,
    },
    articleBody: generateSchemaOrgProperty('article:body'),
    articleSection:
      generateSchemaOrgProperty('article:section') ||
      getCategories(breadcrumbs),
    image:
      generateSchemaOrgProperty('article:image') ||
      generateSchemaOrgProperty('og:image') ||
      image,
    keywords: metadata?.keywords,
  };
};
