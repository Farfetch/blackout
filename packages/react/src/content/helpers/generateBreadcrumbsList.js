/**
 * Generate Structured Data (JSON-LD) for Breadcrumbs.
 *
 * @function generateBreadcrumbsList
 * @memberof module:content/helpers
 *
 * @param {object} breadcrumbs - Breadcrumbs data.
 * @param {string} breadcrumbs.url - Breadcrumbs Item URL.
 * @param {string} breadcrumbs.name - Breadcrumbs Item Name.
 *
 * @returns {object} - JSON-LD Schema.org object for Breadcrumbs.
 */
export default breadcrumbs => ({
  '@context': 'http://schema.org/',
  '@type': 'BreadcrumbList',
  itemListElement: breadcrumbs
    .filter(breadcrumb => (breadcrumb.url || '').length > 0)
    .map((breadcrumb, index) => ({
      '@type': 'ListItem',
      position: index + 1,
      item: {
        '@id': breadcrumb.url,
        name: breadcrumb.name,
      },
    })),
});
