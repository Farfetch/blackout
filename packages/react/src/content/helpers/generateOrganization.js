import * as schemaProperties from '../components/schemas/schemaProperties';
import { getMetatag } from '../utils';

/**
 * Generate Structured Data (JSON-LD) for Organization.
 *
 * @function generateOrganization
 * @memberof module:content/helpers
 *
 * @param {object} metadata - SEO metadata for type Organization on Homepage.
 * @param {object} name - Name of the Organization.
 * @param {string} url - Relative url of the page (location.pathname).
 * @param {string} logoUrl - Complete url for logotype.
 * @param {object} address - Full address information.
 * @param {string} address.street - Stress Address name.
 * @param {string} address.locality - Stress Locality name.
 * @param {string} address.region - Stress Region name.
 * @param {string} address.postalCode - Stress Postal Code name.
 * @param {string} address.country - Stress Country name.
 * @param {object} contact - Full Contact information.
 * @param {string} contact.phone - Phone number contact.
 * @param {string} contact.type - Contact type (e.g. Customer Service).
 * @param {string} contact.address - Contact address name.
 * @param {string} contact.email - E-mail contact.
 * @param {string} contact.option - Contact Option (e.g. A toll-free number or support for hearing-impaired callers).
 * @param {string} contact.areaServed - The geographic area where a service or offered item is provided.
 * @param {Array}  sameAs - SameAs links of organization (e.g. Links to facebook and/or instagram).
 *
 * @returns {object} - JSON-LD Schema.org object for an Organisation.
 */
export default (metadata, name, url, logoUrl, address, contact, sameAs) => {
  // Validate if certain description exist and return the property content.
  const generateSchemaOrgProperty = property =>
    getMetatag(property, metadata?.metatags);

  return {
    '@context': schemaProperties.DEFAULT_CONTEXT,
    '@type': schemaProperties.DEFAULT_ORGANISATION,
    name: generateSchemaOrgProperty('organization:name') || name,
    url: generateSchemaOrgProperty('organization:url') || url,
    logo: generateSchemaOrgProperty('organization:logo') || logoUrl,
    sameAs: generateSchemaOrgProperty('organization:sameAS') || sameAs,
    address: {
      '@context': schemaProperties.DEFAULT_CONTEXT,
      '@type': schemaProperties.DEFAULT_TYPE_ADDRESS,
      streetAddress:
        generateSchemaOrgProperty('organization:streetAddress') ||
        address?.street,
      addressLocality:
        generateSchemaOrgProperty('organization:addressLocality') ||
        address?.locality,
      addressRegion:
        generateSchemaOrgProperty('organization:addressRegion') ||
        address?.region,
      postalCode:
        generateSchemaOrgProperty('organization:postalCode') ||
        address?.postalCode,
      addressCountry:
        generateSchemaOrgProperty('organization:addressCountry') ||
        address?.country,
    },
    contactPoint: {
      '@context': schemaProperties.DEFAULT_CONTEXT,
      '@type': schemaProperties.DEFAULT_TYPE_CONTACT_POINT,
      telephone:
        generateSchemaOrgProperty('organization:telephone') || contact?.phone,
      contactType:
        generateSchemaOrgProperty('organization:contactType') || contact?.type,
      address:
        generateSchemaOrgProperty('organization:address') || contact?.address,
      email: generateSchemaOrgProperty('organization:email') || contact?.email,
      contactOption:
        generateSchemaOrgProperty('organization:contactOption') ||
        contact?.option,
      areaServed:
        generateSchemaOrgProperty('organization:areaServed') ||
        contact?.areaServed,
    },
  };
};
