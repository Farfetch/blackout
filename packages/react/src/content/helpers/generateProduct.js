import * as schemaProperties from '../components/schemas/schemaProperties';
import { getConditionSchemaOrg, getMetatag } from '../utils';

/**
 * Generate Structured Data (JSON-LD) for Product Details.
 *
 * @function generateProduct
 * @memberof module:content/helpers
 *
 * @param {object} product - All details data for the current Product.
 * @param {object} metadata - All SEO metadata for the current Product.
 * @param {string} lastCategory - Last category name.
 * @param {string} url - Relative url of the product (location.pathname).
 * @param {string} seller - Seller name for this particular product.
 *
 * @returns {object} - JSON-LD Schema.org object for Product Details.
 */
export default (product, metadata, lastCategory, url, seller) => {
  // Validate if certain description exist and return the property content.
  const generateSchemaOrgProperty = property =>
    getMetatag(property, metadata?.metatags);

  const metaSeller = generateSchemaOrgProperty('product:seller');

  return {
    '@context': schemaProperties.DEFAULT_CONTEXT,
    '@type': schemaProperties.DEFAULT_TYPE_PRODUCT,
    name:
      generateSchemaOrgProperty('product:name') ||
      generateSchemaOrgProperty('og:title') ||
      product?.name,
    description:
      generateSchemaOrgProperty('product:description') ||
      generateSchemaOrgProperty('og:description') ||
      product?.description,
    image:
      product?.images?.map((image, index) =>
        generateSchemaOrgProperty(`product:image:${index + 1}`)
          ? generateSchemaOrgProperty(`product:image:${index + 1}`)
          : image.url,
      ) || generateSchemaOrgProperty('og:image'),
    sku: generateSchemaOrgProperty('product:sku') || product?.sku,
    productID:
      generateSchemaOrgProperty('product:id') || product?.id?.toString(),
    mpn: generateSchemaOrgProperty('product:mpn') || product?.mpn,
    color:
      generateSchemaOrgProperty('product:color') ||
      product?.colors?.[0].color?.name,
    category: generateSchemaOrgProperty('product:category') || lastCategory,
    brand: {
      '@type': schemaProperties.DEFAULT_BRAND_TYPE,
      name: generateSchemaOrgProperty('product:brand') || product?.brand?.name,
    },
    offers: {
      '@context': schemaProperties.DEFAULT_CONTEXT,
      '@type': schemaProperties.DEFAULT_TYPE_OFFER,
      url: generateSchemaOrgProperty('og:url') || url,
      price:
        generateSchemaOrgProperty('product:price:amount') ||
        product?.price?.includingTaxes,
      priceCurrency:
        generateSchemaOrgProperty('product:price:currency') ||
        product?.currencyIsoCode,
      itemCondition:
        getConditionSchemaOrg(generateSchemaOrgProperty('product:condition')) ||
        schemaProperties.NEW_CONDITION,
      availability: generateSchemaOrgProperty('product:availability')
        ? schemaProperties.IN_STOCK
        : product?.quantity > 0
        ? schemaProperties.IN_STOCK
        : schemaProperties.OUT_OF_STOCK,
      priceValidUntil:
        generateSchemaOrgProperty('product:price:priceValidUntil') ||
        product?.price?.priceValidUntil,
      ...((metaSeller || seller) && {
        seller: {
          '@type': schemaProperties.DEFAULT_ORGANISATION,
          name: metaSeller || seller,
        },
      }),
    },
  };
};
