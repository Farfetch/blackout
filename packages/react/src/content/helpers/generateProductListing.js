import * as schemaProperties from '../components/schemas/schemaProperties';
import { getMetatag } from '../utils';

/**
 * Generate Structured Data (JSON-LD) for Products Listing.
 *
 * @function generateProductListing
 * @memberof module:content/helpers
 *
 * @param {object} listing - All details data for the Products List.
 * @param {object} metadata - All SEO metadata for the Products List.
 * @param {string} url - Relative URL of the page.
 *
 * @returns {object} - JSON-LD Schema.org object for Product Listing.
 */
export default (listing, metadata, url) => {
  const generateSchemaOrgProperty = property =>
    getMetatag(property, metadata?.metatags);

  const generateProductUrl = slug => {
    const initialUrl = url?.substring(0, url?.lastIndexOf('/'));

    return `${initialUrl}/shopping/${slug}`;
  };

  return {
    '@context': schemaProperties.DEFAULT_CONTEXT,
    '@type': schemaProperties.DEFAULT_ITEM_LIST,
    name:
      generateSchemaOrgProperty('productsList:name') ||
      listing?.name ||
      'Products Listing',
    url: url,
    numberOfItems: listing?.products?.totalItems,
    itemListElement: listing?.entries.map((product, index) => ({
      '@type': schemaProperties.DEFAULT_LIST_ITEM,
      position: index + 1,
      name:
        generateSchemaOrgProperty(`productsList:${product.id}:name`) ||
        product.name,
      url: generateProductUrl(product.slug),
    })),
  };
};
