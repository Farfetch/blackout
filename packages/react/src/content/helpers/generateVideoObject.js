import * as schemaProperties from '../components/schemas/schemaProperties';

/**
 * Generate Structured Data (JSON-LD) for Video Objects.
 *
 * @function generateVideoObject
 * @memberof module:content/helpers
 *
 * @param {object} media - All details data for the current Video.
 * @param {string} media.alt - Media file description text.
 * @param {string} media.source - Media source file path.
 * @param {object} media.thumbnails - Media thumbnail images.
 * @param {object} publicationDate - Publication date of the Video.
 *
 * @returns {object} - JSON-LD Schema.org object for Video Objects.
 */
export default (media, publicationDate) => ({
  '@context': schemaProperties.DEFAULT_CONTEXT,
  '@type': schemaProperties.DEFAULT_TYPE_VIDEO,
  name: media?.alt,
  description: media?.alt,
  '@id': media?.source,
  thumbnailUrl: media?.thumbnails?.srcLg,
  embedUrl: media?.source,
  uploadDate: publicationDate,
});
