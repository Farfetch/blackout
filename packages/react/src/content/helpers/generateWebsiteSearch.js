import * as schemaProperties from '../components/schemas/schemaProperties';
import { getMetatag } from '../utils';

/**
 * Generate Structured Data (JSON-LD) for WebsiteSearch.
 *
 * @function generateWebsiteSearch
 * @memberof module:content/helpers
 *
 * @param {object} metadata - SEO metadata for type WebsiteSearch on Homepage.
 * @param {string} url - Relative URL of the page.
 * @param {string} searchTitle - Website title.
 * @param {string} urlTemplate - Complete url to query on.
 *
 * @returns {object} - JSON-LD Schema.org object for Website Search.
 */
export default (metadata, url, searchTitle, urlTemplate) => {
  const generateSchemaOrgProperty = property =>
    getMetatag(property, metadata?.metatags);

  return {
    '@context': schemaProperties.DEFAULT_CONTEXT,
    '@type': schemaProperties.DEFAULT_TYPE_WEBSITE,
    url: generateSchemaOrgProperty('organization:url') || url,
    potentialAction: {
      '@type': schemaProperties.DEFAULT_TYPE_ACTION,
      name: generateSchemaOrgProperty('homepage:searchTitle') || searchTitle,
      target:
        generateSchemaOrgProperty('homepage:searchUrl') ||
        `${urlTemplate}{search_term_string}`,
      'query-input': 'required name=search_term_string',
    },
  };
};
