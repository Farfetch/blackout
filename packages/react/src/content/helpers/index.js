/**
 * Content helpers.
 *
 * @module content/helpers
 * @category Content
 * @subcategory Helpers
 */

export { default as generateArticle } from './generateArticle';
export { default as generateBreadcrumbsList } from './generateBreadcrumbsList';
export { default as generateProduct } from './generateProduct';
export { default as generateProductListing } from './generateProductListing';
export { default as generateOrganization } from './generateOrganization';
export { default as generateVideoObject } from './generateVideoObject';
export { default as generateWebsiteSearch } from './generateWebsiteSearch';
export { default as renderComponent } from './renderComponent';
export { default as renderContent } from './renderContent';
export { default as renderScriptTag } from './renderScriptTag';
