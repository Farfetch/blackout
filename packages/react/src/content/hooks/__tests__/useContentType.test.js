import { cleanup } from '@testing-library/react';
import { ContentType } from './__fixtures__/ContentType.fixtures';
import { doGetContent as doGetContentAction } from '@farfetch/blackout-core/contents/redux';
import {
  expectedNormalizedPayload,
  mockContentsErrorState,
  mockContentsInitialState,
  mockContentsLoadingState,
} from 'tests/__fixtures__/contents';
import { wrap } from '../../../tests/helpers';
import React from 'react';

jest.mock('@farfetch/blackout-core/contents/redux', () => ({
  ...jest.requireActual('@farfetch/blackout-core/contents/redux'),
  doGetContent: jest.fn(() => () => ({ type: 'foo-bar' })),
}));

describe('useContentType', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterEach(cleanup);

  it('should render with the state and selected content type', () => {
    const { container, getByTestId } = wrap(<ContentType codes="career-test" />)
      .withStore(expectedNormalizedPayload)
      .render();

    expect(getByTestId('contentType-title').textContent).toBe(
      'Career Page Title',
    );
    expect(getByTestId('contentType-description').textContent).toBe(
      '<p>Here you can add some text to explain the Career</p>',
    );
    expect(container).toMatchSnapshot();
  });

  it('should render with the state and selected content type and code', () => {
    const { container, getByTestId } = wrap(<ContentType codes="career-test" />)
      .withStore(expectedNormalizedPayload)
      .render();

    expect(getByTestId('contentType-title').textContent).toBe(
      'Career Page Title',
    );
    expect(getByTestId('contentType-description').textContent).toBe(
      '<p>Here you can add some text to explain the Career</p>',
    );
    expect(container).toMatchSnapshot();
  });

  it('should call `fetchContentType` to dispatch a custom content type request', () => {
    const { queryByTestId } = wrap(<ContentType codes="career-test" />)
      .withStore(mockContentsInitialState)
      .render();

    expect(doGetContentAction).toHaveBeenCalledTimes(1);
    expect(queryByTestId('contentType-container')).toBeNull();
  });

  it('should render loading content for a custom content type', () => {
    const { getByTestId } = wrap(<ContentType codes="career-test" />)
      .withStore(mockContentsLoadingState)
      .render();

    expect(getByTestId('contentType-loading').textContent).toBe('Loading...');
  });

  it('should render error content for a custom content type', () => {
    const { getByTestId } = wrap(<ContentType codes="career-test" />)
      .withStore(mockContentsErrorState)
      .render();

    expect(getByTestId('contentType-error').textContent).toBe(
      'Something went wrong!',
    );
  });
});
