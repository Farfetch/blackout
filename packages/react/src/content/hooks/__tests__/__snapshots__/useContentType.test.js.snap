// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useContentType should render with the state and selected content type 1`] = `
<div>
  <div
    data-test="contentType-container"
  >
    <h1
      data-test="contentType-title"
    >
      Career Page Title
    </h1>
    <div
      data-test="contentType-description"
    >
      &lt;p&gt;Here you can add some text to explain the Career&lt;/p&gt;
    </div>
  </div>
</div>
`;

exports[`useContentType should render with the state and selected content type and code 1`] = `
<div>
  <div
    data-test="contentType-container"
  >
    <h1
      data-test="contentType-title"
    >
      Career Page Title
    </h1>
    <div
      data-test="contentType-description"
    >
      &lt;p&gt;Here you can add some text to explain the Career&lt;/p&gt;
    </div>
  </div>
</div>
`;
